### PR TITLE
Webapp > backend: Use one DF per test

### PR DESCRIPTION
**Pull request by AgatheG**
- Great refacto ( [ch74714] )
  - Do not use one df for all the tests
  - Use subclasses for stress tests depending on their type (subpop shift vs feature perturbations)
- Move metrics computation into a dedicated file `metrics.py` (to prepare for the time where the metric used during training will be used instead of hardcoded AUC)
- Remove webapp constants (only one file with constants)
- Probability displayed for the critical samples is now the proper one ( [ch66633] )

### CODE REVIEW

File: python-lib/dku_stress_test_center/metrics.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import numpy as np
+from sklearn.metrics import accuracy_score
+
+def get_performance_metric(metric_name: str):
+    return accuracy_score # TODO
+
+def greater_perf_is_better(metric_name: str):
+    return True # TODO
+
+def worst_group_performance(metric_name: str, y_true: np.array, y_pred: np.array):
```
Comment (simonamaggio) [inline python-lib/dku_stress_test_center/metrics.py:10]: This is correct, but I would make it more general. In the sense that we also give in input the identifiers of the subpopulation each sample belongs to. Could be  `worst_group_performance(metric_name: str, y_true: np.array, y_pred: np.array, subpop_id: np.array)`, so that we can use it also when subpopulations are not defined by the classes. In this specific case, we would have in input a `subpop_id=y_true` (which could be the default, btw), but this is a special case. WDYT?

File: python-lib/dku_stress_test_center/metrics.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+def performance_variation(metric_name: str, clean_y_true: np.array, clean_y_pred: np.array,
+                          perturbed_y_true: np.array, perturbed_y_pred: np.array):
+    greater_is_better = greater_perf_is_better(metric_name)
+    performance_metric = get_performance_metric(metric_name)
+
+    clean_performance_metric = performance_metric(clean_y_true, clean_y_pred)
+    stressed_performance_metric = performance_metric(perturbed_y_true, perturbed_y_pred)
+
+    delta = stressed_performance_metric - clean_performance_metric
+    return delta if greater_is_better else - delta # TODO might be trickier than this
```
Comment (simonamaggio) [inline python-lib/dku_stress_test_center/metrics.py:32]: Yes, I think it can work like this, if we keep the value in their true unit (not a percentage). I'm thinking a MAPE going from 0.01 to 0.05, we display perf drop: -0.04. You will see your MAPE degraded of 4 pts. Makes sense to me.

File: python-lib/dku_stress_test_center/model_accessor.py [hunk 3]
```diff
@@ -44,5 +51,6 @@ def get_per_feature(self):
     def get_predictor(self):
         return self.model_handler.get_predictor()
 
-    def predict(self, df):
-        return self.get_predictor().predict(df)
+    def predict_and_concatenate(self, df, dropna=True):
```
Comment (simonamaggio) [inline python-lib/dku_stress_test_center/model_accessor.py:54]: I don't know why we would have nan predictions and why we need the dropna here. When we drop the rows here, are we sure to drop them in the df then?

File: python-lib/dku_stress_test_center/stress_test_center.py [hunk 4]
```diff
@@ -1,166 +1,177 @@
...
-                        pos_label: str or int):
-    # batch evaluation
-    # return average accuracy drop
-    # return robustness metrics: 1-ASR or imbalanced accuracy for prior shift
+    def compute_metrics(self, perf_metric: str, clean_y_true: np.array, clean_y_pred: np.array,
+                        perturbed_y_true: np.array, perturbed_y_pred: np.array):
+        return {
+            "robustness": stress_resilience(clean_y_pred, perturbed_y_pred),
+            "performance_variation": performance_variation(
+                perf_metric, clean_y_true, clean_y_pred, perturbed_y_true, perturbed_y_pred
```
Comment (simonamaggio) [inline python-lib/dku_stress_test_center/stress_test_center.py:58]: In FeaturePerturbationTest `clean_y_true` and `perturbed_y_true` should always been the same. Don't know if there's anything to do about that, though.

File: python-lib/dku_stress_test_center/stress_test_center.py [hunk 5]
```diff
@@ -1,166 +1,177 @@
...
+        self.tests = defaultdict(list)
+        for test_name, test_config in tests_config.items():
+            test = StressTest.create(test_name, test_config)
+            self.tests[test.TEST_TYPE].append(test)
+
+    def compute_test_metrics(self, test: StressTest):
+        target = self.model_accessor.get_target_variable()
+        df = self.model_accessor.get_original_test_df(sample_fraction=self._sampling_proportion,
+                                                      random_state=self._random_state)
+        perturbed_df = test.perturb_df(df, target)
```
Comment (simonamaggio) [inline python-lib/dku_stress_test_center/stress_test_center.py:107]: For each test, we repeat the sampling step, while we're getting always the same data (thx to the random state). Wouldn't it be better to leave this part out of the `compute_stress_metrics` function ?

Comment (simonamaggio) [reply python-lib/dku_stress_test_center/stress_test_center.py:107]: But I don't have a strong opinion, just wondering if it takes longer without need.

**Review by simonamaggio — CHANGES_REQUESTED** [review]

File: python-lib/dku_stress_test_center/metrics.py (hunk 2 above)
Comment (AgatheG) [reply python-lib/dku_stress_test_center/metrics.py:32]: So from what I understand:

### Classification
- `accuracy`, `precision`, `recall`, `f1`, `cost matrix gain`, `roc auc` & `lift` : **greater_is_better**; use `stressed_performance_metric - clean_performance_metric`
- `log loss` & `calibration loss` : **lower is better**; use `clean_performance_metric - stressed_performance_metric`

### Regression
- `evs` & `r2 score` : **greater is better**
- `mape`,`mae`, `mse`, `rmse`, `rmsle` : **lower is better**
    - [ ] Question: So we can still use `clean_performance_metric - stressed_performance_metric` here? Even when we have formulas involving non linear operators like logs, powers or roots?

:warning: For `correlation` I am not sure. IIUC correlation is between -1 and 1, and the further from 0 the better.
    - [ ] Question:  do we want to use the absolute value of the correlation before doing the subtraction (and then assume greater is better, so we take `stressed_performance_metric - clean_performance_metric`)? Like:
         1. If we have .6 for clean and .3 for perturbed, variation is -.3
         2. If we have - .6 for clean and -.3 for perturbed, variation is also -.3
         3. If we have .6 for clean (resp -.6) and -.3 (resp .3) for perturbed), variation is still -.3

This can be also discussed further on the channel/in a meeting of course.

File: python-lib/dku_stress_test_center/model_accessor.py (hunk 3 above)
Comment (AgatheG) [reply python-lib/dku_stress_test_center/model_accessor.py:54]: 1. We can have empty predictions if one of the features uses `Drop rows (do not predict them either)` as it missing handling method, hence we need the `.dropna(subset=[DkuStressTestCenterConstants.PREDICTION])` on line 55.

2. I am not so sure however why I added the `dropna(how='all')` below 😅 I'll try to see if it can safely be removed

File: python-lib/dku_stress_test_center/metrics.py (hunk 1 above)
Comment (AgatheG) [reply python-lib/dku_stress_test_center/metrics.py:10]: Yep sounds nice indeed!

File: python-lib/dku_stress_test_center/stress_test_center.py (hunk 4 above)
Comment (AgatheG) [reply python-lib/dku_stress_test_center/stress_test_center.py:58]: ooh you're right, I was thinking `predicted_y_true` might be shorter (in the case where missing values are dropped) but actually this should be prevented by `clean_df_with_pred = self._clean_df.loc[test.df_with_pred.index, :]`

File: python-lib/dku_stress_test_center/stress_test_center.py (hunk 5 above)
Comment (AgatheG) [reply python-lib/dku_stress_test_center/stress_test_center.py:107]: Indeed, I suppose it is a memory vs time trade off here, the alternative is to copy eveyrtime the df. Should be slightly faster, we can do it like this

File: python-lib/dku_stress_test_center/metrics.py (hunk 1 above)
Comment (AgatheG) [reply python-lib/dku_stress_test_center/metrics.py:10]: Actually side note, the subpop_id array would still have to be either a subset/superset of the target classes, since we use the array to define the mask `class_mask = y_true == target_class`. Like it must have at least some unique values in common with `y_true` otherwise we do not compute anything

File: python-lib/dku_stress_test_center/metrics.py (hunk 2 above)
Comment (simonamaggio) [reply python-lib/dku_stress_test_center/metrics.py:32]: Classification: all correct
Regression: 
   * yes, I think we should still use the difference, regardless the formula of the metrics
   * yes for correlation, what you propose is the only way it makes sense to me

File: python-lib/dku_stress_test_center/metrics.py (hunk 1 above)
Comment (simonamaggio) [reply python-lib/dku_stress_test_center/metrics.py:10]: So `subpop_id` would contain kind of cluster_ids, independent from the target classes. It represents any partition of the data samples. I think the implementation should be changed to something like:

```
subpop_groups = np.unique(subpop_id)
for subpop_group in subpop_groups:
    group_mask = subpop_id == subpop_group
    performance = get_performance_metric(metric_name)(y_true[group_mask], y_pred[group_mask])
    performances.append(performance)
```

Comment (simonamaggio) [reply python-lib/dku_stress_test_center/metrics.py:10]: when` subpop_id=y_true` we fallback to your implementation is I'm not mistaken.

File: python-lib/dku_stress_test_center/model_accessor.py (hunk 3 above)
Comment (AgatheG) [reply python-lib/dku_stress_test_center/model_accessor.py]: Done in 408cce0

File: python-lib/dku_stress_test_center/stress_test_center.py (hunk 5 above)
Comment (AgatheG) [reply python-lib/dku_stress_test_center/stress_test_center.py]: Done in 39cd1f1

File: python-lib/dku_stress_test_center/metrics.py (hunk 1 above)
Comment (AgatheG) [reply python-lib/dku_stress_test_center/metrics.py]: Done in e507cfe

**Review by simonamaggio — APPROVED** [review]
